runtime/reachable.c

```diff
@@ -0,0 +1,413 @@
+  for (int i = 0; i < elf_info->symtab_num_ents; i++) {
+    sym_ent = &elf_info->symtab[i];
+    if (sym_ent->st_value == sym_unrelocated)
+      return strdup(elf_info->strtab + sym_ent->st_name);
+  }
+
+  return "???";  // not found
+}
+
+static struct {
+  ELF_Uint min;          // minimum number of reachable words
+  char *header_string;  // the header string (per section)
+  int printed_header;   // boolean; 1 = header has been printed
+  int item_num;         // simple counter but not very meaningful
+} print_roots;
+
+// callback routine
+void caml_root_report (value v, value *p /* not used */) {
+  value reachable = Long_val(caml_obj_reachable_words(v));
+  if (reachable >= print_roots.min) {
+    if (!print_roots.printed_header) {
+      printf("\n%s:\n", print_roots.header_string);
+      print_roots.printed_header = 1;
+    }
+    printf("%10" ELF_INT_PRINTF_FORMAT "d item  %3d\n",
+        reachable, print_roots.item_num);
+  }
+  print_roots.item_num++;
+}
+
+extern void print_location(FILE *, struct caml_loc_info * li, int index);
+
+// callback routine
+void caml_stack_report (value v, value *p /* not used */) {
+  if (stack_roots.cur_descr != NULL) {
+    value reachable = Long_val(caml_obj_reachable_words(v));
+    if (reachable >= stack_roots.min) {
+      if (!print_roots.printed_header) {
+        printf("\n");
+        print_roots.printed_header = 1;
+      }
+      if (stack_roots.frame_item == 0) {
+        debuginfo dbi = caml_debuginfo_extract(stack_roots.cur_descr);
+        struct caml_loc_info li;
+        caml_debuginfo_location(dbi, &li);
+        print_location(stdout, &li, stack_roots.frame_item+1);
+        stack_roots.frame_num++;
+      }
+      printf("%10" ELF_INT_PRINTF_FORMAT "d field %3d (frame %d)\n",
+          reachable, stack_roots.frame_item, stack_roots.frame_num-1);
+    }
+  }
+}
+
+void caml_do_stack_roots(long min) {
+  stack_roots.min = min;
+  stack_roots.frame_num = 0;
+  caml_do_local_roots(caml_stack_report, Caml_state->bottom_of_stack,
+                      Caml_state->last_return_address, Caml_state->gc_regs,
+                      Caml_state->local_roots);
+
+}
+
+/* Reports how much memory is reachable from each heap root for those that
+reach at least min words.  This is the main code implementing Gc.print_reachable.
+
+Reads the executable file, which must be in ELF format, to get symbol names
+for global roots.  The symbol names incorporate the source file name.  The
+symbol names for dynamic roots is saved at load time in caml_register_dyn_global.
+
+Accessing roots is modelled on caml_do_roots.
+
+Stack frames are printed only when one of the fields has more than the minimum number
+of words.
+ 
+*/
+void caml_do_print_reachable(long min) {
+  Elf_info elf_info;
+  int printed_glob_header = 0;
+  int printed_dyn_header = 0;
+  int printed_local_C_header = 0;  
+  struct caml__roots_block *lr;
+  dyn_root_link *lnk;
+
+  char *msg = caml_elf_setup(&elf_info);
```

Comment:
`caml_elf_setup` reads the entire executable file into memory, released before `caml_do_print_reachable` returns.  Coq, a fairly large application for OCaml, is only 26 MB, which seems negligible compared to today's DRAM sizes.  It would be an issue only if `malloc` starts to fail.